R-package/src/install.libs.R

```diff
@@ -41,71 +111,65 @@ if (!use_precompile) {
   setwd(build_dir)
 
   # Prepare installation steps
-  cmake_cmd <- "cmake "
-  build_cmd <- "make _lightgbm"
+  cmake_args <- NULL
+  build_cmd <- "make"
+  build_args <- "_lightgbm"
   lib_folder <- file.path(source_dir, fsep = "/")
 
   if (use_gpu) {
-    cmake_cmd <- paste0(cmake_cmd, " -DUSE_GPU=ON ")
+    cmake_args <- c(cmake_args, "-DUSE_GPU=ON")
   }
   if (R_ver >= 3.5) {
-    cmake_cmd <- paste0(cmake_cmd, " -DUSE_R35=ON ")
+    cmake_args <- c(cmake_args, "-DUSE_R35=ON")
   }
-  cmake_cmd <- paste0(cmake_cmd, " -DBUILD_FOR_R=ON ")
+  cmake_args <- c(cmake_args, "-DBUILD_FOR_R=ON")
 
   # Pass in R version, used to help find R executable for linking
   R_version_string <- paste(
     R.Version()[["major"]]
     , R.Version()[["minor"]]
     , sep = "."
   )
-  cmake_cmd <- sprintf(
-    paste0(cmake_cmd, " -DCMAKE_R_VERSION='%s' ")
-    , R_version_string
-  )
+  r_version_arg <- sprintf("-DCMAKE_R_VERSION='%s'", R_version_string)
+  cmake_args <- c(cmake_args, r_version_arg)
+
+  # the checks below might already run `cmake -G`. If they do, set this flag
+  # to TRUE to avoid re-running it later
+  makefiles_already_generated <- FALSE
 
   # Check if Windows installation (for gcc vs Visual Studio)
   if (WINDOWS) {
     if (use_mingw) {
       print("Trying to build with MinGW")
-      cmake_cmd <- paste0(cmake_cmd, " -G \"MinGW Makefiles\" ")
-      build_cmd <- "mingw32-make.exe _lightgbm"
-      system(paste0(cmake_cmd, " ..")) # Must build twice for Windows due sh.exe in Rtools
+      # Must build twice for Windows due sh.exe in Rtools
+      cmake_args <- c(cmake_args, "-G", shQuote("MinGW Makefiles"))
+      .run_shell_command("cmake", c(cmake_args, ".."), strict = FALSE)
```

Comment:
I know the `strict = FALSE` looks a little strange here. I was surprised to learn that right now (on `master`, not specific to this PR), `cmake -G "MinGW Makefiles"` is working but returning a non-zero exit code.

You can see some errors in [the most recent build on `master`](https://ci.appveyor.com/project/guolinke/lightgbm/builds/32728411/job/jpvhc1pkgghirltt)

![image](https://user-images.githubusercontent.com/7608904/81464301-ff697280-9185-11ea-8a88-fd614eaf33e7.png)

I did see that it might be possible to fix this by making sure mingw is on `PATH` (https://stackoverflow.com/a/32922266/3986677), but when I tried a few versions of that it didn't work.

We can safely ignore this for the sake of this PR. It's not a user-facing issue, since the non-0 exit code is handled internally in this R script and since installation succeeds.

I would like to address it eventually so that it doesn't lead to bug reports or waste users' time when they're debugging. I've created #3060

Reply:
@jameslamb Maybe it is because we have 2 instances of MinGW in PATH? One comes from RTools and another one is ordinary MinGW-w64 we use to test Python builds:
https://github.com/microsoft/LightGBM/blob/1b4cddfbb233d8483680cc35ab4385887d4e342f/.appveyor.yml#L22

Reply:
Anyway, I think this line should be executed only for Python builds to prevent weird bugs in the future.

Reply:
Ah interesting! I could try moving that into only Python builds.

Reply:
I just tried this today and I don't think it worked. But I think it's not necessary to fix for t his PR...I'll post my findings on #3060